### PR TITLE
check cmake version before passing the parallel flag

### PR DESCRIPTION
**Pull request by MoAlyousef**
Hello
This PR passes the --version flag to cmake, holds the value in a tuple, then checks whether the version is 3.12 or above.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -607,18 +607,15 @@ impl Config {
                     cmd.arg("-AARM64");
                 } else if target.contains("i686") {
                     use cc::windows_registry::{find_vs_version, VsVers};
-                    match find_vs_version() {
```
Comment (alexcrichton) [inline src/lib.rs:610]: Can unrelated changes like this be split out into separate PRs and/or separate commits?

File: src/lib.rs [hunk 2]
```diff
@@ -973,3 +976,16 @@ fn getenv_unwrap(v: &str) -> String {
 fn fail(s: &str) -> ! {
     panic!("\n{}\n\nbuild script failed, must exit now", s)
 }
+
+// Returns the major, minor and patch versions of cmake
+fn cmake_version(executable: &OsStr) -> io::Result<(u8, u8, u8)> {
+    let cmd = Command::new(executable).arg("--version").output()?;
+    let version = String::from_utf8_lossy(&cmd.stdout);
+    let version: Vec<&str> = version.split_whitespace().collect();
+    let version: Vec<&str> = version[2].split('.').collect();
```
Comment (alexcrichton) [inline src/lib.rs:985]: I think the indexing here should be fallible and I don't think that parse errors below should be papered over with 0.

**Review by alexcrichton — COMMENTED** [review]
Thanks for this, although part of https://github.com/alexcrichton/cmake-rs/pull/123 was actually removing code that ran `make -jN` which means that this is somewhat of a regression where if an older cmake is found then it won't run parallel builds whereas previously before it would.

I don't know of a great way to bridge that though because I'd prefer to not have both systems in place. I suppose not passing `--parallel` is probably the best option for now.

File: src/lib.rs (hunk 1 above)
Comment (MoAlyousef) [reply src/lib.rs]: Removed the lint changes

File: src/lib.rs [hunk 3]
```diff
@@ -808,7 +808,13 @@ impl Config {
                 {
                     cmd.env("MAKEFLAGS", makeflags);
                 }
-                _ => {}
+                _ => {
+                    if let Ok(s) = env::var("NUM_JOBS") {
+                        if !cmake_has_parallel(&executable) {
+                            cmd.env("MAKEFLAGS", format!("-j{}", s));
+                        }
```
Comment (alexcrichton) [inline src/lib.rs:815]: Can this be folded into the processing of `NUM_JOBS` below so the env var and the `cmake --version` command doesn't run twice?

File: src/lib.rs [hunk 4]
```diff
@@ -821,8 +827,9 @@ impl Config {
         cmd.arg("--config").arg(&profile);
 
         if let Ok(s) = env::var("NUM_JOBS") {
-            // See https://cmake.org/cmake/help/v3.12/manual/cmake.1.html#build-tool-mode
-            cmd.arg("--parallel").arg(s);
+            if cmake_has_parallel(&executable) {
+                cmd.arg("--parallel").arg(s);
+            }
```
Comment (alexcrichton) [inline src/lib.rs:832]: Could this add some sort of warning if `--parallel` isn't passed? I can imagine one day the version parsing fails for some reason and having a warning here might help debug a bit.

File: src/lib.rs [hunk 5]
```diff
@@ -973,3 +980,29 @@ fn getenv_unwrap(v: &str) -> String {
 fn fail(s: &str) -> ! {
     panic!("\n{}\n\nbuild script failed, must exit now", s)
 }
+
+// Returns the major, minor and patch versions of cmake
+fn cmake_version(executable: &OsStr) -> Result<(u8, u8, u8), Box<dyn std::error::Error>> {
```
Comment (alexcrichton) [inline src/lib.rs:985]: I don't think this error is ever read, so it's fine to return an `Option<T>` from this function.

File: src/lib.rs [hunk 6]
```diff
@@ -973,3 +980,29 @@ fn getenv_unwrap(v: &str) -> String {
...
+    for i in version.iter() {
+        let curr = i.parse()?;
+        temp.push(curr);
+    }
+    Ok((temp[0], temp[1], temp[2]))
+}
+
+fn cmake_has_parallel(executable: &OsStr) -> bool {
+    if let Ok(cmake_version) = cmake_version(&executable) {
+        if (cmake_version.0 > 3) || (cmake_version.0 == 3 && cmake_version.1 >= 12) {
```
Comment (alexcrichton) [inline src/lib.rs:1000]: I think this would work as `if cmake_version >= (3, 12, 0)`

Comment (MoAlyousef) [reply src/lib.rs]: huh it does work

File: src/lib.rs (hunk 4 above)
Comment (MoAlyousef) [reply src/lib.rs:829]: Added a warning here, and another in the cmake_has_parallel function (I think it might be more appropriate there).

File: src/lib.rs [hunk 7]
```diff
@@ -821,8 +821,12 @@ impl Config {
         cmd.arg("--config").arg(&profile);
 
         if let Ok(s) = env::var("NUM_JOBS") {
-            // See https://cmake.org/cmake/help/v3.12/manual/cmake.1.html#build-tool-mode
-            cmd.arg("--parallel").arg(s);
+            if cmake_has_parallel(&executable) {
+                cmd.arg("--parallel").arg(s);
+            } else {
+                println!("cargo:warning=The parallel flag wasn't passed to CMake!");
```
Comment (alexcrichton) [inline src/lib.rs:827]: Could the be a bit more conditional where this is only printed if `NUM_JOBS` is greater than one, there's no `Makefile`, and could the cmake version be printed here? (noting that the minimum version is 3.12.0)

File: src/lib.rs [hunk 8]
```diff
@@ -808,7 +815,12 @@ impl Config {
                 {
                     cmd.env("MAKEFLAGS", makeflags);
                 }
-                _ => {}
+                _ => {
+                    if !cmake_has_parallel && num_jobs > 1 {
+                        println!("cargo:warning=The parallel flag was not passed to CMake since the detected version is below 3.12!");
```
Comment (alexcrichton) [inline src/lib.rs:820]: If the `-j` flag is passed to `make` then I don't think that this warning should be printed since parallel builds do happen. This should probably only be reserved for builds where parallelism isn't specified.

Comment (alexcrichton) [reply src/lib.rs:820]: I also think it would be helpful to print out the version detected here as well.

File: src/lib.rs [hunk 9]
```diff
@@ -973,3 +984,35 @@ fn getenv_unwrap(v: &str) -> String {
...
+    let cmd = Command::new(executable).arg("--version").output().ok()?;
+    let version = String::from_utf8_lossy(&cmd.stdout);
+    let version: Vec<&str> = version.split_whitespace().collect();
+    let version: Vec<&str> = version[2].split('.').collect();
+    let mut temp = vec![];
+    for i in version.iter() {
+        let curr = i.parse().ok()?;
+        temp.push(curr);
+    }
+    Some((temp[0], temp[1], temp[2]))
```
Comment (alexcrichton) [inline src/lib.rs:999]: I think you'll want to unroll the loop manually because otherwise there's no guarantee that these indexes are actually present and otherwise this could panic.

File: src/lib.rs [hunk 10]
```diff
@@ -973,3 +984,35 @@ fn getenv_unwrap(v: &str) -> String {
...
+// Check if cmake has the --parallel flag by checking against version 3.12
+fn cmake_has_parallel(executable: &OsStr) -> bool {
+    if let Some(cmake_version) = cmake_version(&executable) {
+        if cmake_version >= (3, 12, 0) {
+            true
+        } else {
+            println!(
+                "cargo:warning=CMake version {}.{}.{} detected, No parallel flag will be passed!",
+                cmake_version.0, cmake_version.1, cmake_version.2
+            );
```
Comment (alexcrichton) [inline src/lib.rs:1011]: This warning isn't quite right in the case of makefiles, so I think that this should get deferred until later.

File: src/lib.rs [hunk 11]
```diff
@@ -973,3 +984,35 @@ fn getenv_unwrap(v: &str) -> String {
...
+            false
+        }
+    } else {
+        println!("cargo:warning=cmake-rs might have failed at parsing the version output!");
```
Comment (alexcrichton) [inline src/lib.rs:1015]: I think this warning is a bit misleading, this crate did indeed fail to parse the output instead of "might" have failed to parse, and I think it would be good to mention the output that failed to parse to assist with debugging if necessary.
